### ISSUE TITLE
[RFC]: DSA-CP Decoupling, PCP Migration, and DSA-CP Sunset

### ISSUE THREAD

**Issue by pisceskkk** [vllm-project/vllm-ascend#14002]
## Motivation

This RFC defines the migration from the legacy `enable_dsa_cp` SFA path to Prefill Context Parallelism (PCP) on Model Runner V2 (MRV2).

The migration starts by removing DSA-CP-specific state and branches from `sfa_v1.py`. DSA-CP becomes an explicit SFA specialization in `context_parallel/sfa_cp.py`, while the native SFA metadata builder and implementation remain the common source of truth. This boundary is tracked by [PR #13921](https://github.com/vllm-project/vllm-ascend/pull/13921).

The standalone DSA-CP implementation is transitional. The refactored PCP path will become its user-facing superset: it must preserve the effective TP+SP+DSA-CP capability while adding MRV2 virtual-batch partitioning and causal-attention load balancing. After PCP provides complete and validated SFA model support, `enable_dsa_cp` will be deprecated and the legacy DSA-CP feature and implementation will be removed.

PCP is not an alias for DSA-CP and does not inherit from it. PCP owns balanced prefill work partitioning at the MRV2 input-batch level. SFA/DSA backend code owns sparse-attention-specific tensor layouts and kernel preparation. DCP remains an independent axis and the sole owner of sequence-sharded KV-cache layout.

### Background

The current DSA-CP implementation provides an optimized SFA execution path by combining TP, sequence-parallel hidden-state handling, token/rotary/slot partitioning, KV communication, full-head attention layouts, and specialized `o_proj` handling. Historically, these concerns were embedded as `enable_dsa_cp` branches in the native SFA backend.

That structure has three long-term problems:

1. `sfa_v1.py` is responsible for both the native algorithm and one concrete parallel execution mode.
2. DSA-CP token partitioning is backend-local and cannot reuse the upstream MRV2 virtual-batch lifecycle or its load-balanced causal partition.
3. Keeping DSA-CP and PCP as permanent user features would create two overlapping ways to partition SFA prefill work, with different metadata, graph, speculative decoding, and maintenance paths.

The MRV2 PCP design in [RFC #12622](https://github.com/vllm-project/vllm-ascend/issues/12622) provides the long-term control plane:

- the scheduler retains a global batch;
- the PCP manager builds rank-local virtual rows;
- prefill tokens are assigned using a balanced `2P`-chunk causal partition;
- decode rows are replicated as required by the PCP contract; and
- rank-local results are restored to global token order before logits and sampling.

SFA should consume that lifecycle rather than maintain a second DSA-specific batch partitioner. Decoupling DSA-CP first creates a reviewable boundary and a temporary compatibility path while MRV2 SFA coverage is completed.

### Goals

- Remove `enable_dsa_cp` state, buffers, metadata fields, and conditional execution branches from native `AscendSFAMetadata`, `AscendSFAMetadataBuilder`, and `AscendSFAImpl`.
- Preserve one common SFA metadata and attention forward flow.
- Keep the existing DSA-CP behavior available during the PCP migration.
- Adapt the SFA backend to the upstream-compatible MRV2 PCP virtual-batch lifecycle.
- Preserve the functional baseline currently provided by TP+SP+DSA-CP.
- Add PCP causal-attention load balancing without duplicating the SFA backend.
- Make PCP, TP, SP, and DCP ownership explicit and non-overlapping.
- Complete MRV2 support for SFA models before deprecating `enable_dsa_cp`.
- Sunset the legacy DSA-CP configuration and code only after published correctness, compatibility, and performance gates pass.

### Terminology and ownership

| Term | Responsibility |
| --- | --- |
| TP | Partitions attention heads and tensor-parallel weights |
| SP | Partitions or gathers TP-local hidden states around supported model operations |
| DSA | Sparse-attention algorithm, index selection, and SFA-specific tensor/kernel contract |
| DSA-CP | Legacy SFA execution feature selected by `enable_dsa_cp`; combines DSA-specific token/layout processing with TP/SP communication |
| PCP | MRV2 compute-partitioning axis for balanced prefill execution; expands the process world |
| DCP | Decode KV-cache sequence-sharding axis; reuses ranks and exclusively owns KV slot/shard layout |
| MRV2 | Model Runner V2 and its virtual-batch PCP lifecycle |

The phrase **PCP replaces DSA-CP** means product and user capability replacement. It does not mean that PCP absorbs the DSA sparse-attention algorithm or that the two implementations become one class. The DSA-specific kernel and layout adaptations remain SFA backend responsibilities, but their input partition comes from MRV2 PCP rather than from the legacy `enable_dsa_cp` feature.

The mandatory ownership rules are:

1. PCP owns global-to-rank-local prefill work partitioning and global-order restoration.
2. The SFA backend owns sparse indexer metadata, SFA Q/KV layout, kernel preparation, and SFA output layout.
3. DCP alone owns sequence-sharded KV slots, block scaling, and global-to-local KV index remapping.
4. TP/SP own their existing head, weight, and hidden-state communication.
5. No SFA backend recreates an independent PCP scheduler or virtual-batch manager.

#### Transitional architecture

The decoupling change makes native SFA independent from the concrete CP mode:

```mermaid
classDiagram
    class AscendSFAMetadataBuilder
    class AscendSFAImpl
    class AscendSFADSACPMetadataBuilder
    class AscendSFADSACPImpl
    class AscendSFADCPMetadataBuilder
    class AscendSFADCPImpl
    class AscendSFADSADCPMetadataBuilder
    class AscendSFADSADCPImpl

    AscendSFAMetadataBuilder <|-- AscendSFADSACPMetadataBuilder
    AscendSFAMetadataBuilder <|-- AscendSFADCPMetadataBuilder
    AscendSFADSACPMetadataBuilder <|-- AscendSFADSADCPMetadataBuilder
    AscendSFADCPMetadataBuilder <|-- AscendSFADSADCPMetadataBuilder

    AscendSFAImpl <|-- AscendSFADSACPImpl
    AscendSFAImpl <|-- AscendSFADCPImpl
    AscendSFADSACPImpl <|-- AscendSFADSADCPImpl
    AscendSFADCPImpl <|-- AscendSFADSADCPImpl
```

`sfa_v1.py` owns the native metadata and shared attention flow. `context_parallel/sfa_cp.py` owns DSA-CP, DCP, and their explicit transitional composition. The combined implementation uses cooperative inheritance and does not copy the complete SFA `forward` method.

This structure is an isolation boundary, not the final user-facing PCP design.

#### Target architecture

The target path moves prefill partitioning above the attention backend:

```mermaid
flowchart TB
    G["Global MRV2 InputBatch"]
    P["PCPManager.partition_batch()<br/>balanced virtual rows"]
    M["Rank-local SFA metadata"]
    S["Shared native SFA flow<br/>plus sparse backend hooks"]
    K["KV insertion<br/>DCP ownership only"]
    R["PCP restore index<br/>global token order"]
    O["Logits and sampling"]

    G --> P --> M --> S --> K --> R --> O
```

For PCP size `P`, PCP assigns paired chunks from a `2P`-chunk partition so that early, more expensive causal chunks are balanced with later, cheaper chunks. This replaces the legacy DSA-CP token partition as the user-level prefill distribution mechanism.

The rank-local SFA adapter must still preserve DSA behavior:

- sparse indexer keys and selection metadata follow PCP virtual-token order;
- rotary inputs, slot mappings, query lengths, and key lengths describe the rank-local virtual batch;
- PCP-local prefill K/V fragments are gathered or inserted according to the MRV2 PCP contract;
- decode-only and prefill/mixed output layouts remain distinguishable;
- the TP-sharded `o_proj` parameter remains the persistent source of truth; temporary full-weight buffers are used only where the SFA layout requires them; and
- output restoration returns tokens to global order before logits and sampling.

When DCP is enabled, the flow is orthogonal:

```text
process world size       = TP * PCP * PP * DP
KV-cache shard count     = DCP
scheduler block scaling  = DCP only
prefill compute partition = PCP
```

PCP ranks contribute prefill computation but do not create an additional KV storage shard.

### Rollout

The work proceeds through the ordered TODO list below. The DSA-CP decoupling keeps current behavior while isolating the legacy implementation. PCP/MRV2 SFA adaptation then moves prefill partitioning to the MRV2 virtual-batch lifecycle. After MRV2 fully supports the agreed SFA model and feature scope, the project will deprecate `enable_dsa_cp`, publish a migration window, and remove the legacy feature and code.

### Tracking TODO list

- [ ] **DSA-CP independent decoupling** — [PR #13921](https://github.com/vllm-project/vllm-ascend/pull/13921).
- [ ] **PCP + MRV2 + SFA backend adaptation** — implement the SFA and DSA backend contract defined by [RFC #12622](https://github.com/vllm-project/vllm-ascend/issues/12622).
- [ ] **Complete MRV2 support for SFA models** — close graph, speculative decoding, C8, P/D, cache-sharing, accuracy, and performance gaps.
- [ ] **DSA-CP feature sunset** — deprecate `enable_dsa_cp`, complete the migration window, then remove the feature and legacy code.

The checklist records merged and validated work items. A submitted PR does not by itself complete an item.

### Compatibility and failure policy

- PR #13921 preserves current CLI and additional-config behavior.
- `enable_dsa_cp` remains supported until PCP/MRV2 reaches the agreed SFA replacement scope and the deprecation window is announced.
- PCP does not silently activate when `enable_dsa_cp` is set.
- During migration, unsupported MRV2 PCP+SFA combinations fail during configuration or model-runner initialization.
- PCP must not alter DCP block size, cache capacity, slot ownership, or shard count.
- The legacy path receives correctness and critical bug fixes during the deprecation window, but no new capability expansion after deprecation.
- Removal requires an announced release and a migration guide; it is not an incidental cleanup in another refactor.

## Feedback Period

The feedback period is at least one week from the creation of this RFC issue. It should close only after maintainers agree on:

- PCP as the long-term user-facing replacement for DSA-CP;
- the boundary between MRV2 PCP and SFA/DSA backend responsibilities;
- the ordered TODO list and DSA-CP sunset preconditions;
- the initial MRV2 SFA replacement matrix; and
- the deprecation and removal policy.

## Any Other Things

This RFC covers only the DSA-CP-to-PCP migration and sunset lifecycle. The orthogonal PCP/DCP topology and the broader MRV2 adaptation remain owned by [RFC #12622](https://github.com/vllm-project/vllm-ascend/issues/12622). Follow-up implementation PRs should reference both RFCs where applicable and update the tracking checklist in this issue.

### References

- [vLLM Ascend RFC #12622: Orthogonal PCP and DCP on Model Runner V2](https://github.com/vllm-project/vllm-ascend/issues/12622)
- [vLLM Ascend PR #13921: Decouple DSA-CP and DCP backends](https://github.com/vllm-project/vllm-ascend/pull/13921)
- [vLLM RFC #46358: Unify Context Parallelism in vLLM](https://github.com/vllm-project/vllm/issues/46358)
